tests/python/test_multidevice.py

```diff
@@ -112,3 +113,362 @@ def multidevice_schedule(self):
+                self.ops.permute(self.mlp_linear1_weight, [0, 2, 1]),
+                [d, 1, h * 4 // d, h],
+                [0, 2, 3],
+            ),
+        )
+        T217_matmul = self.ops.sum(T217_local_matmul, [0])
+        T217_biasadd = self.ops.add(
+            T217_matmul,
+            self.ops.broadcast_in_dim(self.mlp_linear1_bias, [1, 1, h], [2]),
+        )
+        T217 = self.ops.cast(T217_biasadd, dtype=DataType.BFloat16)
+        T218 = self.ops.cast(T217, dtype=DataType.Float)
+        T219 = self.ops.cast(T35, dtype=DataType.Float)
+        T220 = self.ops.mul(T218, T219)
+        S221 = self.define_scalar(1.11111, dtype=DataType.Double)
+        T222 = self.ops.mul(T220, S221)
+        T223 = self.ops.add(T158, T222)
+        T224 = self.ops.cast(T223, dtype=DataType.BFloat16)
+        # See the caller of this FusionDefinition for what these outputs mean.
+        self.add_output(T37)
+        self.add_output(T55)
+        self.add_output(T78)
+        self.add_output(T141)
+        self.add_output(T142)
+        self.add_output(T143)
+        self.add_output(T144)
+        self.add_output(T158)
+        self.add_output(T160)
+        self.add_output(T178)
+        self.add_output(T224)
+
+    def multidevice_schedule(self):
+        mesh = self.sched._create_device_mesh(range(self._num_devices))
+        for in_tv in [
+            self.input,
+            self.layernorm0_weight,
+            self.layernorm0_bias,
+            self.mha_linear0_weight,
+            self.mha_linear0_bias,
+            self.mha_linear1_weight,
+            self.mha_linear1_bias,
+            self.layernorm1_weight,
+            self.layernorm1_bias,
+            self.mlp_linear0_weight,
+            self.mlp_linear0_bias,
+            self.mlp_linear1_weight,
+            self.mlp_linear1_bias,
+        ]:
+            self.sched._set_device_mesh(in_tv, mesh)
+
+        self.sched.parallelize(self.mlp_linear0_weight, 0, nvfuser.ParallelType.mesh_x)
+        self.sched.parallelize(self.mlp_linear0_bias, 0, nvfuser.ParallelType.mesh_x)
+        self.sched.parallelize(self.mlp_linear1_weight, 0, nvfuser.ParallelType.mesh_x)
+
+
+@pytest.mark.skipif(
+    utils.is_pre_ampere(),
+    reason="Flash Attention is only supported on Ampere and newer devices.",
+)
+@pytest.mark.mpi
+def test_transformer_forward(mpi_test):
+    d = mpi_test.size
+    rank = mpi_test.rank
+
+    b, s, a, h = 2, 2048, 96, 12288
+
+    if h * 4 % d != 0:
+        pytest.skip(
+            f"We only support even split, so {h} * 4 has to be divisible by {d}."
+        )
+
+    torch.cuda.set_device(mpi_test.local_rank)
+
+    mlp_linear0_weight = torch.randn(
+        d, h * 4 // d, h, dtype=torch.bfloat16, device="cuda"
+    )
+    mlp_linear0_bias = torch.randn(d, h * 4 // d, dtype=torch.bfloat16, device="cuda")
+    mlp_linear1_weight = torch.randn(
+        d, h, h * 4 // d, dtype=torch.bfloat16, device="cuda"
+    )
+    ins = [
+        29,
```

Comment:
can you add with inlined comment what each input corresponds to?

Reply:
ditto

Reply:
is it really the same consideration as in https://github.com/NVIDIA/Fuser/pull/3045/files#r1790935495 ?

Reply:
I added a code comment to point readers to their meanings. I could add inlined comments for each argument, but I would have a hard time making sure they stay consistent with those in the definition. So I prefer adding a pointer so there's only one source of truth.

Reply:
thanks!